### PR TITLE
Do not buffer the request body by default when reading forms.

### PR DESCRIPTION
**Pull request by Tratcher**
#578 @rynowak @muratg @lodejard 
We still have to buffer multipart file bodies, but we can disable buffering for the full request body.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Http/Features/HttpResponseFeature.cs [hunk 1]
```diff
@@ -25,17 +25,17 @@ public HttpResponseFeature()
...
         }
 
-        public void OnStarting(Func<object, Task> callback, object state)
+        public virtual void OnStarting(Func<object, Task> callback, object state)
         {
             throw new NotImplementedException();
         }
 
-        public void OnCompleted(Func<object, Task> callback, object state)
+        public virtual void OnCompleted(Func<object, Task> callback, object state)
```
Comment (muratg) [inline src/Microsoft.AspNetCore.Http/Features/HttpResponseFeature.cs]: virtual for testing?

File: test/Microsoft.AspNetCore.Http.Tests/FormFeatureTests.cs [hunk 2]
```diff
@@ -13,68 +13,90 @@ namespace Microsoft.AspNetCore.Http.Features.Internal
...
             Assert.Null(formFeature);
 
-            // Act
             var formCollection = await context.Request.ReadFormAsync();
 
-            // Assert
             Assert.Equal("bar", formCollection["foo"]);
             Assert.Equal("2", formCollection["baz"]);
-            Assert.Equal(0, context.Request.Body.Position);
-            Assert.True(context.Request.Body.CanSeek);
```
Comment (muratg) [inline test/Microsoft.AspNetCore.Http.Tests/FormFeatureTests.cs]: Do we need to test these two elsewhere?

Comment (muratg) [inline test/Microsoft.AspNetCore.Http.Tests/FormFeatureTests.cs]: Or have something like `if (bufferRequest) // still do the asserts`?

**Comment by muratg** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Http/Features/HttpResponseFeature.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Http/Features/HttpResponseFeature.cs]: Yes, it's simpler than re-implementing the whole interface.

**Pull request merged**